### PR TITLE
small change in copyWithoutReturn to handle CCoerce

### PR DESCRIPTION
**Pull request by RenaudFondeur**
Suppress cCoerce at the top of return expression when we suppress return in inlining.

cCoerce can be found in returns to make them match their method return type.
During inlining we can create a copy of parts of the AST without their return. If there is a cCoerce at the top of their expression, it may be no longer needed and can produce a 'warning: expression result unused'.

### CODE REVIEW

File: smalltalksrc/Slang/TReturnNode.class.st [hunk 1]
```diff
@@ -107,7 +107,10 @@ TReturnNode >> children [
 
 { #category : 'transformations' }
 TReturnNode >> copyWithoutReturn [
-	^expression
+
+	(expression isSend and: [ expression selector beginsWith: #cCoerce ])
+		ifTrue: [ ^ expression arguments first ].
```
Comment (PalumboN) [inline smalltalksrc/Slang/TReturnNode.class.st]: Would be nice to add a comment about why of this, for the future devs.

**Review by PalumboN — COMMENTED** [review]
🆒 !

I suggest to add a comment and **any test**. 
I don't remember well, but part of the inlinings are tested. Would be nice to add one example of this.

Also, I did not get why the compiler throws `warning: expression result unused` due a cast 🤔  (maybe the test helps me to understand).

File: smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st
Comment (PalumboN) [inline smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st:1]: 🤩

File: smalltalksrc/Slang/TMethod.class.st [hunk 2]
```diff
@@ -1678,6 +1679,7 @@ TMethod >> inlineSend: aSendNode directReturn: directReturn exitVar: exitVar in:
 				 in: aCodeGen);
 		addAll: meth statements. "method body"
 	directReturn ifTrue: [
+		meth parseTree parent: aSendNode parent.
```
Comment (PalumboN) [inline smalltalksrc/Slang/TMethod.class.st:1682]: 🤔  Is this ok? The parent of a send is not the method? So the parent of `meth` is another method?

File: smalltalksrc/Slang/TParseNode.class.st [hunk 3]
```diff
@@ -283,6 +283,15 @@ TParseNode >> isNonNullCCode [
...
 ]
 
+{ #category : 'testing' }
+TParseNode >> isNotExpression [
+
+	^ parent isNil or: [
+		  parent isTMethod or: [
+			  parent isSend not and: [
+				  parent isReturn not and: [ parent isAssignment not ] ] ] ]
+]
```
Comment (PalumboN) [inline smalltalksrc/Slang/TParseNode.class.st:293]: I don't understand completely this:

1. The method is _negated_ (is**Not**Expression). It is more difficult to follow a negation than an affirmation. Would be nice to have an affirmative method and do the negation in the use: `self isExpression not`.
2. This definition has sense for all `TParseNode`? Or only for `TReturnNode` (who call this method).
3. I don't follow the condition anyway:
  - Why if the paren is nil is not an expression?
  - I understand that the only _non-expression_ nodes are returns, assignments, (maybe if / case ?). Why is not asking that and is always asking for the parent?

File: smalltalksrc/Slang/TStatementListNode.class.st [hunk 4]
```diff
@@ -356,7 +356,9 @@ TStatementListNode >> copyWithoutReturn [
 
 	| newStmtsListNode |
 	self assert: self endsWithReturn.
-	statements size = 1 ifTrue: [ ^ statements last expression ].
+	statements size = 1 ifTrue: [
+		statements last parent: self parent.
```
Comment (PalumboN) [inline smalltalksrc/Slang/TStatementListNode.class.st:360]: Why this?

**Review by PalumboN — COMMENTED** [review]
Cooollll! This is better, mainly because of the tests 💯 🚀 🚦 

I left new comments / questions.

File: smalltalksrc/Slang/TMethod.class.st (hunk 2 above)
Comment (guillep) [reply smalltalksrc/Slang/TMethod.class.st:1682]: Let's open an issue for this

File: smalltalksrc/Slang/TParseNode.class.st (hunk 3 above)
Comment (guillep) [reply smalltalksrc/Slang/TParseNode.class.st:293]: Let's open an issue for this

File: smalltalksrc/Slang/TStatementListNode.class.st (hunk 4 above)
Comment (guillep) [reply smalltalksrc/Slang/TStatementListNode.class.st:360]: I will reference this in the first issue I created

**Pull request merged**